tempesta_fw/classifier/frang.c

```diff
@@ -502,14 +546,70 @@ frang_http_host_check(const TfwHttpReq *req, FrangAcc *ra)
 	return ret;
 }
 
+static unsigned short
+frang_resp_next_idx(unsigned short ab, const unsigned short len)
+{
+	return ++ab >= len ? 0 : ab;
+}
+
+static void
+frang_resp_keep_frame(FrangRespCodeStat *stat, const unsigned long tc)
```

Comment:
As i see, the `frang_req_limit()` has pretty the same meaning and mostly the same should be done in your patch. Look, in the patch you may recount current rate of responses (with given status code) twice: first during response processing, then during processing the next request. I suggest solely increment number of responses for _current part_ of current time frame during response processing and evaluate limit overflow during request processing. 

With this approach you don't need scale factor and variable-sized `FrangRespCodeStat.buf`, and structure like `FrangAcc.FrangRates` will cover all your needs. The only difference is sampling rate: `FrangAcc.FrangRates` is sampled `FRANG_FREQ` times per second (since limits are per-second) while your data can be sampled `FRANG_FREQ` times per `FrangHttpRespCodeBlock.tf` (since limit is per user defined time frame).

Reply:
>  suggest solely increment number of responses for current part of current time frame during response processing and evaluate limit overflow during request processing.

We discussed this with @bartolootrit . The problem is that we limit number of bad responses in _time_ _frame_. Consider we had a limit for 10 responses per 10 seconds and a user generates 2 bad responses per second leading to blocking at 6th second. Next the user stops sending and we have to update the statistic on next request - we must allow 2 more bad responses on 7th second. So it seems we have to recalculate the statistic either on requests and responses.